File: objective-c/oauthdemoapp/ServerManager.m
//
//  ServerManager.m
//  oauthdemoapp
//

#import "ServerManager.h"
#import "LoginViewController.h"
#import "Token.h"
#import "HttpHelper.h"

@interface ServerManager ()

// @property (strong, nonatomic) Token *token;

@end

@implementation ServerManager

// parameters for authorization request
static NSString *authURL        = @"https://sequencing.com/oauth2/authorize";
static NSString *redirect_uri   = @"authapp://Default/Authcallback";
static NSString *response_type  = @"code";
static NSString *state          = @"statecodecallbackmd5";
static NSString *clientID       = @"oAuth2 Demo ObjectiveC";
static NSString *scope          = @"demo";

// parameters for token request
static NSString *tokenURL       = @"https://sequencing.com/oauth2/token";
static NSString *client_secret  = @"RZw8FcGerU9e1hvS5E-iuMb8j8Qa9cxI-0vfXnVRGaMvMT3TcvJme-Pnmr635IoE434KXAjelp47BcWsCrhk0g";
static NSString *grant_type     = @"authorization_code";
static NSString *redirect_uritk = @"https://objectivec-oauth-demo.sequencing.com/Default/Authcallback";

// parameters for demo data request
static NSString *apiURL         = @"https://api.sequencing.com";
static NSString *demoPath       = @"/DataSourceList?sample=true";


+ (ServerManager *) sharedInstance {
    static ServerManager *manager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[ServerManager alloc] init];
    });
    return manager;
}

- (id)init {
    self = [super init];
    if (self) {
        //self.token = [Token new];
    }
    return self;
}

- (void)authorizeUser:(void(^)(NSArray *data))completion {
    
    NSString *client_id = [clientID stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLQueryAllowedCharacterSet];
    NSString *urlString = [NSString stringWithFormat:
                           @"%@?"
                           "redirect_uri=%@&"
                           "response_type=%@&"
                           "state=%@&"
                           "client_id=%@&"
                           "scope=%@",
                           authURL, redirect_uri, response_type, state, client_id, scope];
    NSURL *url = [NSURL URLWithString:urlString];
    
    NSLog(@"===== 1st request (authorize user) =====");
    
    LoginViewController *loginViewController =
    [[LoginViewController alloc] initWithURL:(NSURL *)url andCompletionBlock:^(NSMutableDictionary *response) {
        if (response) {
            // NSLog(@"response %@", response);
            
            // first, must check if "state" from response matches "state" in request
            if (![[response objectForKey:@"state"] isEqualToString:state]) {
                NSLog(@"state mismatch, response is being spoofed");
                if (completion) {
                    completion(nil);
                }
            } else {
                
                // state matches - we can proceed with token request
                NSLog(@"===== 2nd request (getting token) =====");
    
                [self postForTokenWithCode:[response objectForKey:@"code"] onSuccess:^(Token *token) {
                    if (token) {
                        
                        NSLog(@"=== 3rd request (getting demo data) ===");
                        [self getForDemoDataWithToken:token onSuccess:^(NSArray *demoData) {
                            // NSLog(@"demodata: %@", demoData);
                            if (completion) {
                                completion(demoData);
                            }
                        } onFailure:^(NSError *error) {
                            NSLog(@"error: %@", error);
                            if (completion) {
                                completion(nil);
                            }
                        }];
                    } else {
                        if (completion) {
                            completion(nil);
                        }
                    }
                } onFailure:^(NSError *error) {
                    NSLog(@"error = %@", [error localizedDescription]);
                    if (completion) {
                        completion(nil);
                    }
                }];
            }
        } else if (completion) {
            completion(nil);
        }
    }];
    
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:loginViewController];
    UIViewController *mainVC = [[[[UIApplication sharedApplication] windows] firstObject] rootViewController];
    [mainVC presentViewController:nav
                         animated:YES
                       completion:nil];
}

- (void)postForTokenWithCode:(NSString *)code
                   onSuccess:(void(^)(Token *token))success
                   onFailure:(void(^)(NSError *error))failure {

    NSDictionary *postParameters = [NSDictionary dictionaryWithObjectsAndKeys:
                            grant_type,     @"grant_type",
                            code,           @"code",
                            redirect_uri,   @"redirect_uri", nil];
    
    [HttpHelper execHttpRequestWithUrl:tokenURL
                             andMethod:@"POST"
                            andHeaders:nil
                           andUsername:clientID
                           andPassword:client_secret
                              andToken:nil
                          andAuthScope:@"Basic"
                         andParameters:postParameters
                            andHandler:^(NSString* responseText, NSURLResponse* response, NSError* error) {
        if (response) {
            // NSLog(@"response text: %@", responseText);
            NSError *jsonError;
            NSData *jsonData = [responseText dataUsingEncoding:NSUTF8StringEncoding];
            NSDictionary *parsedObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                         options:0
                                                                           error:&jsonError];
            if (jsonError != nil) {
                // NSLog(@"Error: %@", jsonError);
                if (success) {
                    success(nil);
                }
            } else {
                // NSLog(@"%@", parsedObject);
                Token *token = [Token new];
                token.accessToken = [parsedObject objectForKey:@"access_token"];
                NSTimeInterval interval = [[parsedObject objectForKey:@"access_token"] doubleValue];
                token.expirationDate = [NSDate dateWithTimeIntervalSinceNow:interval];
                token.tokenType = [parsedObject objectForKey:@"token_type"];
                token.scope = [parsedObject objectForKey:@"scope"];
                token.refreshToken = [parsedObject objectForKey:@"refresh_token"];
                if (success) {
                    success(token);
                }
            }
        } else if (failure) {
            failure(error);
        }
    }];
}

- (void)getForDemoDataWithToken:(Token *)token
                      onSuccess:(void(^)(NSArray *demoData))success
                      onFailure:(void(^)(NSError *error))failure {
    
    NSString *apiUrlForDemo = [[NSString alloc] initWithFormat:@"%@%@", apiURL, demoPath];
    
    [HttpHelper execHttpRequestWithUrl:apiUrlForDemo
                             andMethod:@"GET"
                            andHeaders:nil
                           andUsername:nil
                           andPassword:nil
                              andToken:token.accessToken
                          andAuthScope:@"Bearer"
                         andParameters:nil
                            andHandler:^(NSString* responseText, NSURLResponse* response, NSError* error) {
                                
                                if (response) {
                                    
                                    NSError *jsonError;
                                    NSData *jsonData = [responseText dataUsingEncoding:NSUTF8StringEncoding];
                                    NSArray *parsedObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                                                 options:0
                                                                                                   error:&jsonError];
                                    if (jsonError != nil) {
                                        NSLog(@"Error: %@", jsonError);
                                        if (success) {
                                            success(nil);
                                        }
                                    } else {
                                        if (success) {
                                            success(parsedObject);
                                        }
                                    }
                                } else if (failure) {
                                    failure(error);
                                }
                            }];
}



@end

File: swift/oauthdemoapp-swift/ServerManager.swift
//
//  ServerManager.swift
//

import UIKit
import Foundation

class ServerManager: NSObject {
    
    static let sharedInstance = ServerManager() //designated initializer
    
    // parameters for authorization request
    let authURL: String         = "https://sequencing.com/oauth2/authorize"
    let redirect_uri: String    = "authapp://Default/Authcallback"
    let response_type: String   = "code"
    let state: String           = "statecodecallbackmd5"
    let clientID: String        = "oAuth2 Demo ObjectiveC"
    let scope: String           = "demo"
    
    // parameters for token request
    let tokenURL: String        = "https://sequencing.com/oauth2/token"
    let client_secret: String   = "RZw8FcGerU9e1hvS5E-iuMb8j8Qa9cxI-0vfXnVRGaMvMT3TcvJme-Pnmr635IoE434KXAjelp47BcWsCrhk0g"
    let grant_type: String      = "authorization_code"
    
    // parameters for demo data request
    let apiURL: String          = "https://api.sequencing.com"
    let demoPath: String        = "/DataSourceList?sample=true"
    
    // activity indicator with label
    var messageFrame = UIView()
    var activityIndicator = UIActivityIndicatorView()
    var strLabel = UILabel()
    let mainVC = UIApplication.sharedApplication().windows.first!.rootViewController
    
    let kMainQueue = dispatch_get_main_queue()
    
    
    // MARK: Functions
    func authorizeUser(completion: (demoData: NSArray?) -> Void) -> Void {
        self.activityIndicator = UIActivityIndicatorView.init(activityIndicatorStyle: UIActivityIndicatorViewStyle.Gray)
        
        let client_id: String = clientID.stringByAddingPercentEncodingWithAllowedCharacters(.URLQueryAllowedCharacterSet())!
        let urlString = authURL + "?" +
            "redirect_uri=" + redirect_uri + "&" +
            "response_type=" + response_type + "&" +
            "state=" + state + "&" +
            "client_id=" + client_id + "&" +
            "scope=" + scope
        let urlRequest: NSURL = NSURL(string: urlString)!
        
        print("===== 1st request (authorize user) =====")
        let loginViewController: LoginViewController = LoginViewController.init(url: urlRequest) { (response) -> Void in
            if response != nil {
                // first, must check if "state" from response matches "state" in request
                let stateInResponse = response?.objectForKey("state") as! String
                if stateInResponse != self.state {
                    print("state mismatch, response is being spoofed")
                    completion(demoData: nil)
                } else {
                    
                    // state matches - we can proceed with token request
                    self.startActivityIndicator()
                    print("===== 2nd request (getting token) =====")
                    let codeFromResponse = response?.objectForKey("code") as! String
                    self.postForTokenWithCode(codeFromResponse, completion: { (token, error) -> Void in
                        if token != nil {
                            // print(token?.accessToken)
                            
                            print("===== 3rd request (getting demo data) =====")
                            self.getForDemoDataWithToken(token!, completion: { (demoData, error) -> Void in
                                if demoData != nil {
                                    self.stopActivityIndicator()
                                    completion(demoData: demoData)
                                    
                                } else  if error != nil {
                                    print(error)
                                    self.stopActivityIndicator()
                                    completion(demoData: nil)
                                }
                            })
                            
                        } else if error != nil {
                            print(error)
                            self.stopActivityIndicator()
                            completion(demoData: nil)
                        }
                    })
                }
            } else {
                self.stopActivityIndicator()
                completion(demoData: nil)
            }
        }
        let nav = UINavigationController(rootViewController: loginViewController)
        let mainVC = UIApplication.sharedApplication().windows.first!.rootViewController
        mainVC!.presentViewController(nav, animated: true, completion: nil)
    }
    
    
    func postForTokenWithCode(code: String, completion: (token: Token?, error: NSError?) -> Void) -> Void {
        let grant_typeString: String = "" + self.grant_type
        let redirect_uriString: String = "" + self.redirect_uri
        let postParameters: [String: String] = ["grant_type": grant_typeString, "code": code, "redirect_uri": redirect_uriString]
        
        HttpHelper.sharedInstance.execHttpRequestWithUrl(
            tokenURL,
            method: "POST",
            headers: nil,
            username: clientID,
            password: client_secret,
            token: nil,
            authScope: "Basic",
            parameters: postParameters) { (responseText, response, error) -> Void in
                if responseText != nil {
                    let jsonData = responseText?.dataUsingEncoding(NSUTF8StringEncoding)
                    do {
                        if let jsonParsed = try NSJSONSerialization.JSONObjectWithData(jsonData!, options: []) as? NSDictionary {
                            let token = Token()
                            
                            if let value = jsonParsed.objectForKey("access_token") as! String? {
                                token.accessToken = value
                            }
                            if let value = jsonParsed.objectForKey("expires_in") as! String? {
                                let myDouble: NSTimeInterval = Double(value)!
                                token.expirationDate = NSDate(timeIntervalSinceNow: myDouble)
                            }
                            if let value = jsonParsed.objectForKey("token_type") as! String? {
                                token.tokenType = value
                            }
                            if let value = jsonParsed.objectForKey("scope") as! String? {
                                token.scope = value
                            }
                            if let value = jsonParsed.objectForKey("refresh_token") as! String? {
                                token.refreshToken = value
                            }
                            completion(token: token, error: nil)
                        }
                    } catch let error as NSError {
                        print("json error" + error.localizedDescription)
                    }
                } else if error != nil {
                    print("error: ")
                    print(error)
                    print("response: ")
                    print(response)
                    completion(token: nil, error: error)
                }
        }
        
    }
    
    func getForDemoDataWithToken(token: Token, completion: (demoData: NSArray?/*Array<String>?*/, error: NSError?) -> Void) -> Void {
        let apiUrlForDemo: String = self.apiURL + self.demoPath
        HttpHelper.sharedInstance.execHttpRequestWithUrl(
            apiUrlForDemo,
            method: "GET",
            headers: nil,
            username: nil,
            password: nil,
            token: token.accessToken,
            authScope: "Bearer",
            parameters: nil) { (responseText, response, error) -> Void in
                if responseText != nil {
                    let jsonData = responseText?.dataUsingEncoding(NSUTF8StringEncoding)
                    do {
                        if let jsonParsed = try NSJSONSerialization.JSONObjectWithData(jsonData!, options: []) as? NSArray {
                            completion(demoData: jsonParsed, error: nil)
                        }
                    } catch let error as NSError {
                        print("json error" + error.localizedDescription)
                    }
                    
                } else if error != nil {
                    print("error: ")
                    print(error)
                    print("response: ")
                    print(response)
                    completion(demoData: nil, error: error)
                }
                
        }
    }
    
    // MARK: Activity indicator
    func startActivityIndicator() -> Void {
        dispatch_async(kMainQueue) { () -> Void in
            if self.mainVC != nil {
                self.strLabel = UILabel(frame: CGRect(x: 50, y: 0, width: 150, height: 50))
                self.strLabel.text = "Getting demo data"
                self.strLabel.textColor = UIColor.grayColor()
                
                self.messageFrame = UIView(frame: CGRect(x: self.mainVC!.view.frame.midX - 100, y: self.mainVC!.view.frame.midY - 25 , width: 250, height: 40))
                self.messageFrame.layer.cornerRadius = 15
                self.messageFrame.backgroundColor = UIColor.clearColor()
                
                self.activityIndicator = UIActivityIndicatorView(activityIndicatorStyle: UIActivityIndicatorViewStyle.Gray)
                self.activityIndicator.frame = CGRect(x: 0, y: 0, width: 50, height: 50)
                self.activityIndicator.startAnimating()
                
                self.messageFrame.addSubview(self.activityIndicator)
                self.messageFrame.addSubview(self.strLabel)
                self.mainVC!.view.addSubview(self.messageFrame)
            }
        }
    }
    
    func stopActivityIndicator() -> Void {
        dispatch_async(kMainQueue) { () -> Void in
            self.activityIndicator.stopAnimating()
            self.messageFrame.removeFromSuperview()
        }
    }

}
